src/main/java/de/tum/in/www1/artemis/repository/specs/PostSpecs.java

```diff
@@ -0,0 +1,156 @@
+
+                Predicate answered = criteriaBuilder.equal(joinedAnswers.get(AnswerPost_.AUTHOR), userId);
+                Predicate reacted = criteriaBuilder.equal(joinedReactions.get(Reaction_.USER), userId);
+
+                return criteriaBuilder.or(answered, reacted);
+            }
+        });
+    }
+
+    public static Specification<Post> getUnresolvedSpecification(Boolean unresolved) {
+        return ((root, query, criteriaBuilder) -> {
+            if (unresolved == null || !unresolved) {
+                return null;
+            }
+            else {
+                root.join(Post_.ANSWERS, JoinType.LEFT);
+
+                Predicate postsWithoutCourseWideContext = criteriaBuilder.isNull(root.get(Post_.COURSE_WIDE_CONTEXT));
+                Predicate notAnnouncementPosts = criteriaBuilder.notEqual(root.get(Post_.COURSE_WIDE_CONTEXT), CourseWideContext.ANNOUNCEMENT);
+
+                // Post should not have any answer that resolves
+                Subquery<Long> subQuery = query.subquery(Long.class);
+                Root<AnswerPost> subRoot = subQuery.from(AnswerPost.class);
+                Predicate subPredicate = criteriaBuilder.equal(subRoot.get(AnswerPost_.RESOLVES_POST), Boolean.TRUE);
+                Predicate postBinder = criteriaBuilder.equal(root.get(Post_.ID), subRoot.get(AnswerPost_.POST).get(Post_.ID));
+                subQuery.select(subRoot.get(AnswerPost_.ID)).where(criteriaBuilder.and(postBinder, subPredicate));
+
+                Predicate notAnnouncement = criteriaBuilder.or(postsWithoutCourseWideContext, notAnnouncementPosts);
+                Predicate notResolves = criteriaBuilder.exists(subQuery).not();
+
+                return criteriaBuilder.and(notAnnouncement, notResolves);
+            }
+        });
+    }
+
+    public static Specification<Post> getSortSpecification(PostSortCriterion postSortCriterion, SortingOrder sortingOrder) {
+        return ((root, query, criteriaBuilder) -> {
+
+            List<Order> orderList = new ArrayList<>();
+
+            Expression<Object> caseExpression = criteriaBuilder.selectCase()
+                    .when(criteriaBuilder.equal(root.get(Post_.DISPLAY_PRIORITY), criteriaBuilder.literal(DisplayPriority.PINNED)), 1)
+                    .when(criteriaBuilder.equal(root.get(Post_.DISPLAY_PRIORITY), criteriaBuilder.literal(DisplayPriority.NONE)), 2)
+                    .when(criteriaBuilder.equal(root.get(Post_.DISPLAY_PRIORITY), criteriaBuilder.literal(DisplayPriority.ARCHIVED)), 3);
+            orderList.add(criteriaBuilder.asc(caseExpression));
+
+            if (postSortCriterion != null && postSortCriterion != PostSortCriterion.CREATION_DATE) {
+                orderList.add(sortingOrder == SortingOrder.ASCENDING ? criteriaBuilder.asc(root.get(Post_.CREATION_DATE)) : criteriaBuilder.desc(root.get(Post_.CREATION_DATE)));
+            }
+
+            query.orderBy(orderList);
```

Comment:
Can this be done as we did in `UserRepository.java -> getAllManagedUsers` :thinking:

e.g. similar to:
```
var sorting = Sort.by(userSearch.getSortedColumn());
sorting = userSearch.getSortingOrder() == SortingOrder.ASCENDING ? sorting.ascending() : sorting.descending();
final var sorted = PageRequest.of(userSearch.getPage(), userSearch.getPageSize(), sorting);

...

return findAll(..., sorted);
```

Reply:
This is just a draft. In fact I've not completed the sorting part yet. 😆

Reply:
No unfortunately. We have a lot of variations. We can't just sort by column. In some cases we have to sort by number of answers or number of a specific type of emoji. I'll take a look at the sorting part after I'm done with the filters though. This part very preliminary. (It is just a draft at the moment 🙄 )